repository/src/main/resources/alfresco/ibatis/org.alfresco.repo.domain.dialect.Dialect/node-common-SqlMap.xml

```diff
@@ -1147,6 +1147,21 @@
             parentNode.id = #{parentNode.id} and
             a.child_node_id IS NULL
     </select>
+
+    <select id="select_RootGroups" parameterType="ChildAssoc" resultType="java.lang.String">
+        select distinct z.qname_localname
+            from alf_child_assoc a
+            join alf_child_assoc z
+            on z.parent_node_id = #{parentNode.id}
+        where z.qname_localname not in
+        (
+            select distinct  a.qname_localname
+            from alf_child_assoc a
+            join alf_child_assoc z
+            on z.parent_node_id = #{parentNode.id}
+            and z.child_node_id = a.parent_node_id
+        )

```

Comment:
If I'm reading this correctly, we're trying to get all the _"orphan"_ parent rows (rows for the which there is no existing child identifying such row as its parent?).

Assuming this is the intended logic to apply here and I interpreted the query correctly, how would the current subquery fare against a `not exists ( select 1 ...` equivalent clause instead? I assume relying on `not exists` might make the intention of the query clearer + it might perform better since it doesn't have to actually fetch any values but just return a boolean.

Also, does the subquery really need to repeat the self-join, or could it be simplified to something like e.g.:

`where not exists
(
    select 1
    from alf_child_assoc aca
    where aca.parent_node_id = z.child_node_id
)`

-, while yielding the same results?

I may be missing some context here 🤔

Reply:
Thanks for the suggestion, after some tweaking, using the `not exists` instead of `not in` improved the query response time by more than 50% in my tests.